core/signeddata.go

```diff
@@ -178,19 +183,54 @@ type VersionedSignedBeaconBlock struct {
 	eth2spec.VersionedSignedBeaconBlock // Could subtype instead of embed, but aligning with Attestation that cannot subtype.
 }
 
-func (b VersionedSignedBeaconBlock) MessageRoot() ([32]byte, error) {
+func (b VersionedSignedBeaconBlock) HashRoot() ([32]byte, error) {
+	roots, err := b.MessageRoots()
+	if err != nil {
+		return [32]byte{}, err
+	}
+
+	// TODO(xenowits/deneb): Return hash tree root of all roots when len(roots) > 1
+	return roots[0], nil
+}
+
+func (b VersionedSignedBeaconBlock) MessageRoots() ([][32]byte, error) {
 	switch b.Version {
 	// No block nil checks since `NewVersionedSignedBeaconBlock` assumed.
 	case eth2spec.DataVersionPhase0:
-		return b.Phase0.Message.HashTreeRoot()
+		root, err := b.Phase0.Message.HashTreeRoot()
+		if err != nil {
+			return nil, errors.Wrap(err, "hash signed block message", z.Str("version", eth2spec.DataVersionPhase0.String()))
+		}
+
+		return [][32]byte{root}, nil
 	case eth2spec.DataVersionAltair:
-		return b.Altair.Message.HashTreeRoot()
+		root, err := b.Altair.Message.HashTreeRoot()
+		if err != nil {
+			return nil, errors.Wrap(err, "hash signed block message", z.Str("version", eth2spec.DataVersionAltair.String()))
+		}
+
+		return [][32]byte{root}, nil
 	case eth2spec.DataVersionBellatrix:
-		return b.Bellatrix.Message.HashTreeRoot()
+		root, err := b.Bellatrix.Message.HashTreeRoot()
+		if err != nil {
+			return nil, errors.Wrap(err, "hash signed block message", z.Str("version", eth2spec.DataVersionBellatrix.String()))
+		}
+
+		return [][32]byte{root}, nil
 	case eth2spec.DataVersionCapella:
-		return b.Capella.Message.HashTreeRoot()
+		root, err := b.Capella.Message.HashTreeRoot()
+		if err != nil {
+			return nil, errors.Wrap(err, "hash signed block message", z.Str("version", eth2spec.DataVersionCapella.String()))
+		}
+
+		return [][32]byte{root}, nil
 	case eth2spec.DataVersionDeneb:
-		return b.Deneb.Message.HashTreeRoot()
+		root, err := b.Deneb.Message.HashTreeRoot()
```

Comment:
This code path will return block root and sidecars root in the Deneb support PR, right?

Reply:
Yeah, absolutely. Something like this:
```go
...
	case eth2spec.DataVersionDeneb:
		var msgRoots [][32]byte
		blockRoot, err := v.Deneb.SignedBlock.Message.HashTreeRoot()
		if err != nil {
			return nil, errors.Wrap(err, "hash signed block message")
		}

		msgRoots = append(msgRoots, blockRoot)

		for _, blob := range v.Deneb.SignedBlobSidecars {
			blobRoot, err := blob.Message.HashTreeRoot()
			if err != nil {
				return nil, errors.Wrap(err, "hash blob sidecar")
			}

			msgRoots = append(msgRoots, blobRoot)
		}

		return msgRoots, nil
...
```